piet-coregraphics/src/gradient.rs

```diff
@@ -0,0 +1,151 @@
+    context::CGContextRef,
+    geometry::CGPoint,
+};
+
+use piet::kurbo::Point;
+use piet::{Color, FixedGradient, FixedLinearGradient, FixedRadialGradient, GradientStop};
+
+// core-graphics does not provide a CGGradient type
+pub enum CGGradientT {}
+pub type CGGradientRef = *mut CGGradientT;
+
+declare_TCFType! {
+    CGGradient, CGGradientRef
+}
+
+impl_TCFType!(CGGradient, CGGradientRef, CGGradientGetTypeID);
+
+/// A wrapper around CGGradient
+#[derive(Clone)]
+pub struct Gradient {
+    cg_grad: CGGradient,
+    piet_grad: FixedGradient,
+}
+
+impl Gradient {
+    pub(crate) fn from_piet_gradient(gradient: FixedGradient) -> Gradient {
+        let cg_grad = match &gradient {
+            FixedGradient::Linear(grad) => new_cg_gradient(&grad.stops),
+            FixedGradient::Radial(grad) => new_cg_gradient(&grad.stops),
+        };
+        Gradient {
+            cg_grad,
+            piet_grad: gradient,
+        }
+    }
+
+    pub(crate) fn first_color(&self) -> Color {
+        match &self.piet_grad {
+            FixedGradient::Linear(grad) => grad.stops.first().map(|g| g.color.clone()),
+            FixedGradient::Radial(grad) => grad.stops.first().map(|g| g.color.clone()),
+        }
+        .unwrap_or(Color::BLACK)
+    }
+
+    pub(crate) fn fill(&self, ctx: &mut CGContextRef) {
+        let context_ref: *mut u8 = ctx as *mut CGContextRef as *mut u8;
+        match self.piet_grad {
+            FixedGradient::Radial(FixedRadialGradient {
+                center,
+                origin_offset,
+                radius,
+                ..
+            }) => {
+                let start_center = to_cgpoint(center + origin_offset);
+                let center = to_cgpoint(center);
+                unsafe {
+                    CGContextDrawRadialGradient(
+                        context_ref,
+                        self.cg_grad.as_concrete_TypeRef(),
+                        start_center,
+                        0.0,
+                        center,
+                        radius as CGFloat,
+                        0,
+                    )
+                }
+            }
+            FixedGradient::Linear(FixedLinearGradient { start, end, .. }) => {
+                let start = to_cgpoint(start);
+                let end = to_cgpoint(end);
+                unsafe {
+                    CGContextDrawLinearGradient(
+                        context_ref,
+                        self.cg_grad.as_concrete_TypeRef(),
+                        start,
+                        end,
+                        0,
+                    )
+                }
+            }
+        }
+    }
+}
+
+fn new_cg_gradient(stops: &[GradientStop]) -> CGGradient {
+    unsafe {
+        //FIXME: is this expensive enough we should be reusing it?
+        let space = CGColorSpace::create_with_name(kCGColorSpaceSRGB).unwrap();
+        let space_ref: *const u8 = &*space as *const CGColorSpaceRef as *const u8;
```

Comment:
This is a mess. Let me suggest the following. Delete this line. Several more related suggestions to follow.